### PR TITLE
Traefik logging and grafana-dashboard integrations

### PR DESCRIPTION
**Pull request by PietroPasotti**
This PR adds loki-logging and grafana-dashboarding integrations to the traefik charm.

## Testing Instructions
deploy cos-lite, then refresh the traefik charm with one built from these sources. Visit the grafana frontend to see the traefik dashboard with the logs.

image

## Release Notes
- Added loki-logging and grafana-dashboarding integrations.

### CODE REVIEW

**Comment by Abuelodelanada** [comment]
> Should we add a loki alert? Seems like a good start could be a general purpose alert for when there's an ERROR level entry in the logs.

It's in another PR: https://github.com/canonical/traefik-k8s-operator/pull/120

**Review by Abuelodelanada — APPROVED** [review]

**Comment by PietroPasotti** [comment]
> * Drop down "Juju Model" is title-cased unlike the others.
Fixed

> * Drop downs "Job" and "Interval": what are they used for?
Unsure, they came with the dashboard (https://grafana.com/grafana/dashboards/4475-traefik/)

> * Pie chart's title is missing
Fixed

> * Average response time: should we add to the title the time range? E.g. "Average response time (last 5 min)". Unless it's averaging across something else than time?
Looking at the formula it looks like it's time. And the graph is showing the last hour, so I added that. But perhaps it was meant to be used together with that 'interval' dropdown somehow?

> * Used sockets: should we set the type of the y axis to integers?
As soon as there's data the values automatically go to integers, but I'll try to hardcode it

**Comment by sed-i** [comment]
> Drop downs "Job" and "Interval": what are they used for?
> Unsure, they came with the dashboard (https://grafana.com/grafana/dashboards/4475-traefik/)

:shrug: Any chance it's redundant given juju topology?

> * Looking at the formula it looks like it's time. And the graph is showing the last hour, so I added that. But perhaps it was meant to be used together with that 'interval' dropdown somehow?

Maybe you're right: the expression doesn't specify duration:
```
"expr": "sum(traefik_backend_request_duration_seconds_sum{protocol=~\"$protocol\"}) / sum(traefik_entrypoint_requests_total{protocol=~\"$protocol\"}) * 1000"
```
but it also doesn't use `$__interval`. Interesting. Perhaps `"intervalFactor": 2` has anything to do with it?

File: src/grafana_dashboards/traefik.json.tmpl [hunk 1]
```diff
@@ -0,0 +1,1454 @@
...
+        "datasource": {
+          "type": "prometheus",
+          "uid": "${prometheusds}"
+        },
+        "definition": "label_values(up{juju_model=\"$juju_model\",juju_model_uuid=\"$juju_model_uuid\",juju_application=\"$juju_application\"},juju_unit)",
+        "hide": 0,
+        "includeAll": false,
+        "label": "juju_unit",
+        "multi": false,
+        "name": "juju_unit",
```
Comment (lucabello) [inline src/grafana_dashboards/traefik.json.tmpl:1322]: Minor pet peeve, I would change this variable display name to `Juju unit` to match the style of `Juju application` and `Juju model`
```suggestion
        "name": "Juju unit",
```

File: src/grafana_dashboards/traefik.json.tmpl [hunk 2]
```diff
@@ -0,0 +1,1454 @@
...
+          "expr": "sum(traefik_entrypoint_open_connections) by (method)",
+          "format": "time_series",
+          "intervalFactor": 1,
+          "legendFormat": "",
+          "refId": "A"
+        }
+      ],
+      "thresholds": [],
+      "timeRegions": [],
+      "title": "ENTRYPOINT - Open Connections",
```
Comment (lucabello) [inline src/grafana_dashboards/traefik.json.tmpl:862]: Another minor pet peeve, me not liking full caps
```suggestion
      "title": "Entrypoint - Open Connections",
```

File: src/grafana_dashboards/traefik.json.tmpl [hunk 3]
```diff
@@ -0,0 +1,1454 @@
...
+          "expr": "sum(traefik_backend_open_connections) by (method)",
+          "format": "time_series",
+          "intervalFactor": 1,
+          "legendFormat": "",
+          "refId": "A"
+        }
+      ],
+      "thresholds": [],
+      "timeRegions": [],
+      "title": "BACKEND - Open Connections",
```
Comment (lucabello) [inline src/grafana_dashboards/traefik.json.tmpl:957]: Another minor pet peeve
```suggestion
      "title": "Backend - Open Connections",
```

**Review by lucabello — APPROVED** [review]
I have just a couple minor aesthetic remarks, feel free to ignore them :)

File: src/grafana_dashboards/traefik.json.tmpl [hunk 4]
```diff
@@ -0,0 +1,1454 @@
...
+        "name": "prometheusds",
+        "options": [],
+        "query": "prometheus",
+        "refresh": 1,
+        "regex": "",
+        "skipUrlSync": false,
+        "type": "datasource"
+      },
+      {
+        "allValue": "traefik",
```
Comment (rbarry82) [inline src/grafana_dashboards/traefik.json.tmpl:1307]: Technically speaking, all of these dropdowns aren't necessary and will be injected by the dashboard library, but they're not hurting anything here either.

**Review by rbarry82 — APPROVED** [review]

File: src/charm.py [hunk 5]
```diff
@@ -866,7 +881,8 @@ def _restart_traefik(self):
                 _TRAEFIK_SERVICE_NAME: {
                     "override": "replace",
                     "summary": "Traefik",
-                    "command": BIN_PATH,
+                    # trick to drop the logs to a file but also keep them available in the pod logs
+                    "command": '/bin/sh -c "{} | tee {}"'.format(BIN_PATH, self._log_path),
```
Comment (dstathis) [inline src/charm.py:885]: +1 This is great

**Review by dstathis — APPROVED** [review]

File: src/charm.py (hunk 5 above)
Comment (PietroPasotti) [reply src/charm.py:885]: I don't remember who I ripped it off of; it's not flour from my own bag (*italian idiom)

**Pull request merged**